docker/developers/docker-compose.mysql.yml

```diff
@@ -0,0 +1,41 @@
+include:
+  - path: ./docker-compose.networks.yml
+
+services:
+  # Docker-managed database
+  mysql-base: &mysql-base
+    hostname: db
+    image: mysql/mysql-server:8.0
+    # You can add default options via the 'command':
+    #command:
+    # - '--log-error-verbosity=3'
+    environment:
+      MYSQL_DATABASE: 'db'
+      MYSQL_USER: 'user'
+      MYSQL_PASSWORD: 'password'
+      MYSQL_ROOT_PASSWORD: 'password'
+      MYSQL_ROOT_HOST: '%'
+    volumes:
+      - mysql_data:/var/lib/mysql
+    # Many default ulimits are just too high
+    # And, so, it will consume way too much memory (16GB+)
+    ulimits:
+      nofile:
+        soft: 65536
+        hard: 65536
+
+  mysql-contained:
+    <<: *mysql-base
+    expose:
+      - "3306"
+    networks:
+      cdo_network_test:
+      cdo_network:
+
+  mysql:
+    <<: *mysql-base
```

Comment:
YAML Merge keys and anchors are a very new concept to me - would it be worth some comments to explain this magic or link to docs? These symbols aren't especially googleable.

Reply:
I'm having some trouble knowing where such documentation should go. This is mostly just an educational thing related to informing anybody reading these how this piece of YAML trivia about how this non-standard implementation works. That documentation would be as such:

This uses YAML anchor expansion syntax, which is supported by Docker Compose, to help keep duplicated content cleanly in one place within the Docker Compose files. With this, you can name blocks of YAML like this:

```
my-service: &my-service-base
  hostname: foo
```

And then use it to pre-populate the block with the contents of that prior block before your additional items:

```
my-service-hosted:
  <<: *my-service-base
  network: host
```

This service will have the properties of its named block and then the additional fields. Any fields that exist in both will be overwritten when read top-down, so in the prior case, if we also specify `hostname` at the end, it will override it. If we specify it as the first line, it will be overwritten by the expanded properties instead.